### DISCUSSION TITLE
Per host configuration

### DISCUSSION THREAD

**Discussion by aurora-xa** [dustinlyons/nixos-config#68] [Q&A] [answered]
Would you mind expanding on how to implement per host configurations? 
I took a look at your personal config repo but I do not follow how you select which 'host' you are installing for. 
Does it just 'work' i.e. hostname = which config it will pick ? 
Many thanks

**Comment by dustinlyons** [comment, marked as answer]
Hi,

In designing this configuration, I made a tradeoff to not create per host configuration as I felt, at the time, reading other configurations was difficult as they were per host with many small module files.

So I wanted to make something more obvious, with Mac-specific and NixOS-specific code broken out into their own directories, filenames like `packages,` `files,` `homebrew,` etc, to denote clearly what's going on. Optimize for learnability over a more canonical configuration.

So it's currently limited to one machine of each platform type. I have a home lab server at home, but instead of adopting this configuration, I'll likely just maintain of fork of this repository for it.

Here's how the code works:

```
darwinConfigurations = nixpkgs.lib.genAttrs darwinSystems (system:
        darwin.lib.darwinSystem {
          inherit system;
          specialArgs = inputs;
          modules = [
            home-manager.darwinModules.home-manager
            nix-homebrew.darwinModules.nix-homebrew
            {
              nix-homebrew = {
                inherit user;
                enable = true;
                taps = {
                  "homebrew/homebrew-core" = homebrew-core;
                  "homebrew/homebrew-cask" = homebrew-cask;
                  "homebrew/homebrew-bundle" = homebrew-bundle;
                };
                mutableTaps = false;
                autoMigrate = true;
              };
            }
            ./hosts/darwin
          ];
        }
      );
```

This line:

```
darwinConfigurations = nixpkgs.lib.genAttrs darwinSystems (system:
```

says

Run the `[genAttrs](https://noogle.dev/f/lib/genAttrs)` function, with a list of strings and a function as arguments. 

Here is our list of strings, `darwinSystems`:

```
darwinSystems = [ "aarch64-darwin" "x86_64-darwin" ];
```

So the `genAttrs` function maps over these two strings, and creates a key value pair where the key is the string and the value is the result of the function, which in our case is a `darwin.lib.darwinSystem`.

So `darwinConfigurations` ends up as:

```
darwinConfigurations = 
        "aarch64-darwin" = darwin.lib.darwinSystem {
          inherit system;
          specialArgs = inputs;
          modules = [
            home-manager.darwinModules.home-manager
            nix-homebrew.darwinModules.nix-homebrew
            {
              nix-homebrew = {
                inherit user;
                enable = true;
                taps = {
                  "homebrew/homebrew-core" = homebrew-core;
                  "homebrew/homebrew-cask" = homebrew-cask;
                  "homebrew/homebrew-bundle" = homebrew-bundle;
                };
                mutableTaps = false;
                autoMigrate = true;
              };
            }
            ./hosts/darwin
          ];
        }
      );

      "x86_64-darwin" = darwin.lib.darwinSystem {
        inherit system;
        specialArgs = inputs;
        modules = [
          home-manager.darwinModules.home-manager
          nix-homebrew.darwinModules.nix-homebrew
          {
            nix-homebrew = {
              inherit user;
              enable = true;
              taps = {
                "homebrew/homebrew-core" = homebrew-core;
                "homebrew/homebrew-cask" = homebrew-cask;
                "homebrew/homebrew-bundle" = homebrew-bundle;
              };
              mutableTaps = false;
              autoMigrate = true;
            };
          }
          ./hosts/darwin
        ];
      }
    );
```

This makes two hosts available by their platform name. When you run the build process, the [scripts look for the platform by name](https://github.com/dustinlyons/nixos-config/blob/10024e13845aeaee58a7999fb701c826bebf88f0/apps/aarch64-darwin/build#L9).

Hope this helps!

**Comment by aurora-xa** [comment]
Thank you for the explanation - I will try and tweak!
